### PR TITLE
(CDAP-6833) Made the privilege fetcher in RemoteOpsService a proxy, t…

### PR DESCRIPTION
**Pull request by bdmogal**
…hat simply proxies list privilege requests to the master, which can communicate with authorization backends.
- DatasetOpExecutor which runs the proxy, now also maintains an authorization cache inside PrivilegeFetcherProxyService, which is similar to the authorization enforcement cache used by program containers, except that it updates itself by fetching privileges via the master.
- Refactored DefaultAuthorizationEnforcementService to move non-enforcement code into an AbstractAuthorizationService class, which is also used by the PrivilegesFetcherProxyService
- Refactored Remote client and handlers for fetching privileges to move common code to abstract classes, which are extended by:
  - RemotePrivilegesFetcherHandler: runs inside master, communicates with authorization backends directly
  - RemotePrivilegesFetcherProxyHandler: runs inside RemoteSystemOpsService, proxies requests from RemotePrivilegesFetcher to RemotePrivilegesFetcherHandler, via PrivilegesFetcherProxyClient
- (CDAP-6639) Updates to authorization cache in the master now do not go via the RemoteSystemOpsService, since master can communicate with Authorization backends directly
  - Fixed MasterAuthenticationContext to get the user id from UserGroupInformation, when requests are made to authorization backends periodically for updating master's authorization cache. Without this, the authentication context would expect SecurityRequestContext to be set, which is null when requests do not go via the Router.
- Most logic of this change is in AuthorizationEnforcementModule.

Jira: 
- [CDAP-6833](https://issues.cask.co/browse/CDAP-6833)
- [CDAP-6639](https://issues.cask.co/browse/CDAP-6639)

Build: http://builds.cask.co/browse/CDAP-DUT4577

### CODE REVIEW

**Comment by bdmogal** [comment]
PR looks large, but is mostly moving code to abstract classes to re-use in two different places: program/service (explore/stream) containers and RemoteSystemOpsService (which runs the proxy).

Most logic is captured in AuthorizationEnforcementModule.

This still needs thorough testing on multi-node clusters. I'm opening the PR to get feedback on the approach.

**Comment by bdmogal** [comment]
Forgot to mention:
The background for this change is that Apache Sentry does not support proxy authentication. It does not issue delegation tokens, like some other Hadoop services. [Relevant thread](http://mail-archives.apache.org/mod_mbox/sentry-dev/201607.mbox/%3CCA%2BDi8eCdDtWzyURCob3HN5q%2B1QqYzgVn5s7uDyniZ9Vk1Wd5Cw%40mail.gmail.com%3E). Although no one has replied to this thread, I've confirmed this with a Sentry committer.

There's also a TODO to this PR: [CDAP-6855](https://issues.cask.co/browse/CDAP-6855)

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/AppFabricTestHelper.java [hunk 1]
```diff
@@ -147,7 +148,11 @@ public static void ensureNamespaceExists(Id.Namespace namespace) throws Exceptio
   public static void ensureNamespaceExists(Id.Namespace namespace, CConfiguration cConf) throws Exception {
     NamespaceAdmin namespaceAdmin = getInjector(cConf).getInstance(NamespaceAdmin.class);
     if (!namespaceAdmin.exists(namespace)) {
-      namespaceAdmin.create(new NamespaceMeta.Builder().setName(namespace).build());
+      try {
+        namespaceAdmin.create(new NamespaceMeta.Builder().setName(namespace).build());
+      } catch (NamespaceAlreadyExistsException e) {
```
Comment (anwar6953) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/AppFabricTestHelper.java:153]: this is just to protect against race condition?
What motivated you to put this here? were you running into the race condition?

File: cdap-security/src/main/java/co/cask/cdap/security/auth/context/MasterAuthenticationContext.java [hunk 2]
```diff
@@ -19,17 +19,40 @@
...
   @Override
   public Principal getPrincipal() {
-    return SecurityRequestContext.toPrincipal();
+    // When requests come in via rest endpoints, the userId is updated inside SecurityRequestContext, so give that
+    // precedence.
+    String userId = SecurityRequestContext.getUserId();
+    // This userId can be null, when the master itself is asynchoronously updating the policy cache, since
+    // during that process the router will not set the SecurityRequestContext. In that case, obtain the userId from
+    // the UserGroupInformation, which will be the user that the master is running as.
+    if (userId == null) {
```
Comment (anwar6953) [inline cdap-security/src/main/java/co/cask/cdap/security/auth/context/MasterAuthenticationContext.java:49]: why is this change needed now?
Why didn't we need to fall back to UGI#getCurrentUser before?

File: cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationService.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+  /**
+   * On an authorization-enabled cluster, if caching is enabled too, updates the cache in the
+   * {@link AuthorizationEnforcementService} with the privileges of the user running the program.
+   */
+  private void updatePrivilegesOfCurrentUser() {
+    String userName;
+    try {
+      userName = UserGroupInformation.getCurrentUser().getShortUserName();
+    } catch (IOException e) {
+      LOG.warn("Error while determining the currently logged in user. Skipping pre-population of authorization cache.",
```
Comment (anwar6953) [inline cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationService.java:187]: btw why isn't this fatal?

File: cdap-security/src/main/java/co/cask/cdap/security/auth/context/MasterAuthenticationContext.java (hunk 2 above)
Comment (bdmogal) [inline cdap-security/src/main/java/co/cask/cdap/security/auth/context/MasterAuthenticationContext.java:49]: When RemoteSystemOperationsService makes a call to AppFabric, SecurityRequestCOntext is not set, because it doesn't go via router. So have to use it. However, gave precedence to SecurityRequestContext so that in master REST APIs, we use SecurityRequestContext.

File: cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationService.java (hunk 3 above)
Comment (bdmogal) [inline cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationService.java:187]: This could be a transient error. Its an optimization to load privileges before starting the program. So its just a warning. If it continues to fail after the program starts, it will become fatal.

Comment (bdmogal) [inline cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationService.java:187]: In the sense that the program will fail.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/AppFabricTestHelper.java (hunk 1 above)
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/AppFabricTestHelper.java:153]: no not a race condition. But the purpose of this method is to "ensure that a namespace exists". So if it already exists, it should not throw, but just ignore.

File: cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationService.java (hunk 3 above)
Comment (anwar6953) [inline cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationService.java:187]: makes sense

**Comment by anwar6953** [comment]
LGTM.
I tested this branch and see that issue still plagues authorization: https://issues.cask.co/browse/CDAP-6871
This is likely a smaller fix, and can come in a separate PR.

**Comment by bdmogal** [comment]
Merging this, since the approach works. Will create follow-up PRs for stream and explore.
